src/Features/Core/Portable/DocumentationComments/CodeFixes/AbstractAddDocCommentNodesCodeFixProvider.cs

```diff
@@ -0,0 +1,173 @@
+        protected abstract string GetValueFromNameAttribute(TXmlNameAttributeSyntax attribute);
+        protected abstract SyntaxNode TryGetDocCommentNode(SyntaxTriviaList parameter);
+        protected abstract string GetXmlElementLocalName(TXmlElementSyntax element);
+        protected abstract List<string> GetParameterNames(TMemberDeclarationSyntax method);
+        protected abstract TXmlElementSyntax GetNewNode(string parameterName, bool isFirstNodeInComment);
+
+        protected async Task<Document> AddParamTagAsync(
+            Document document, TextSpan span, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var parentMethod = root.FindNode(span).FirstAncestorOrSelf<TMemberDeclarationSyntax>();
+            var docCommentNode = TryGetDocCommentNode(parentMethod.GetLeadingTrivia());
+
+            var newDocComment = docCommentNode;
+            var parameterNames = GetParameterNames(parentMethod);
+
+            for (var index = 0; index < parameterNames.Count; index++)
+            {
+                var parameterName = parameterNames[index];
+                var paramNodes = GetParameterNodes(newDocComment, NodeName);
+                if (NodeExists(paramNodes, parameterName))
+                {
+                    continue;
+                }
+
+                var paramsBeforeCurrentParam = parameterNames.TakeWhile(t => t != parameterName).ToList();
+                var paramsAfterCurrentParam = parameterNames.Except(paramsBeforeCurrentParam).ToList();
+                paramsAfterCurrentParam.Remove(parameterName);
+
+                // If the index is not `0`, there is a node before the current one for sure
+                // If the index is `0`, try to add the node after the `summary` node,
+                // only if any existing nodes are at the top level--this route will not
+                // be taken if the existing node is nested in another node
+                var summaryNode = GetParameterNodes(newDocComment, "summary").FirstOrDefault();
+                if (index != 0 || !paramNodes.Any() && summaryNode != null)
+                {
+                    // First, try to get the node before the param node so we know where to insert the new node
+                    SyntaxNode nodeBeforeNewParamNode =  GetLastParamNodeCorrespondingToParameterInList(paramNodes, paramsBeforeCurrentParam) ??
+                                    summaryNode;
+
+                    newDocComment = newDocComment.InsertNodesAfter(nodeBeforeNewParamNode, new[] { GetNewNode(parameterName, false) });
+                    continue;
+                }
+                
+                // At this point, the node has to go at the beginning of the comment
+                var nodeAfterNewParamNode = GetLastParamNodeCorrespondingToParameterInList(paramNodes.AsEnumerable().Reverse(), paramsAfterCurrentParam) ??
```

Comment:
If I pass a reversed list to the method, I can reuse the same method to get the list of parameter names before/after the current name.  I don't need to reverse the list, but I need the names given in reverse.